metaflow/datatools/s3.py

```diff
@@ -449,7 +471,13 @@ def get_root_from_config(cls, echo, create_on_absent=True):
         return DATATOOLS_S3ROOT
 
     def __init__(
-        self, tmproot=".", bucket=None, prefix=None, run=None, s3root=None, **kwargs
+        self,
+        tmproot: str = ".",
+        bucket: Optional[str] = None,
+        prefix: Optional[str] = None,
+        run: Optional[Union[FlowSpec, TRun]] = None,
```

Comment:
Been staring at this for a moment. Why the use of a TypeVar? I know these can be used for generics to enhance autocompletion when using subclasses.

Example:

```python
from typing import Generic, TypeVar
from dataclasses import dataclass
from .some_file import Weapon

TWeapon = TypeVar("TWeapon", bound=Weapon)

@dataclass
class Enemy(Generic[TWeapon]):
    holding: TWeapon

enemy = Enemy(holding=...some weapon subclass instance)
enemy.holding # gives autocompletion for the specific subclass of Weapon in holding
```

Is that somehow what's going on with `TRun`?  Does `Union[type, type]` simply require TypeVar objects rather than the string representation, i.e. `"Run"`?

Reply:
Let me try with "Run". I have to check it works on 3.5 too. IIRC that may be why I went this route but may have been delusional :) .